aea/helpers/dialogue/base.py

```diff
@@ -412,90 +395,145 @@ def last_message(self) -> Optional[Message]:
         for message in list_of_all_messages:
-            if message.message_id == message_id_to_find:
-                result = message
-                break
+            if message.message_id == message_id:
+                return message
 
-        return result
+    def has_message(self, message: Message) -> bool:
+        """
+        Check whether a message exists in this dialogue.
 
-    @property
-    def is_empty(self) -> bool:
+        :param message: the message
+        :return: True if message exists in this dialogue, False otherwise
         """
-        Check whether the dialogue is empty.
+        if self.is_empty:
+            return False
 
-        :return: True if empty, False otherwise
+        if not self.has_message_id(message.message_id):
+            return False
+
+        retrieved_message = self.get_message(message.message_id)
+        return message == retrieved_message
+
+    def has_message_id(self, message_id: int) -> bool:
         """
-        return len(self._outgoing_messages) == 0 and len(self._incoming_messages) == 0
+        Check whether a message with the supplied message id exists in this dialogue.
 
-    def update(self, message: Message) -> bool:
+        :param message_id: the message id
+        :return: True if message with that id exists in this dialogue, False otherwise
+        """
+        if self.is_empty:
+            return False
+
+        return 1 <= message_id <= self.last_message.message_id
+
+    def update(self, message: Message) -> None:
         """
         Extend the list of incoming/outgoing messages with 'message', if 'message' belongs to dialogue and is valid.
 
         :param message: a message to be added
         :return: True if message successfully added, false otherwise
         """
-        if not message.is_incoming:
+        if not message.has_sender:
             message.sender = self.agent_address
 
-        self.ensure_counterparty(message)
-
         if not self.is_belonging_to_dialogue(message):
-            return False
-
-        is_extendable = self.is_valid_next_message(message)
-        if is_extendable:
-            if message.is_incoming:
-                self._incoming_messages.extend([message])
-            else:
-                self._outgoing_messages.extend([message])
-        return is_extendable
+            raise InvalidDialogueMessage(
+                "The message {} does not belong to this dialogue."
+                "The dialogue reference of the message is {}, while the dialogue reference of the dialogue is {}".format(
+                    message.message_id,
+                    message.dialogue_reference,
+                    self.dialogue_label.dialogue_reference,
+                )
+            )
 
-    def ensure_counterparty(self, message: Message) -> None:
-        """
-        Ensure the counterparty is set (set if not) correctly.
+        is_valid_result, is_valid_message = self.is_valid_next_message(message)
```

Comment:
The second variable should not start with `is_` as it is a string; maybe call it `validation_msg`

Reply:
They were meant to be read as the bool result returned by `is_valid` and the message returned by `is_valid`.

Anyway changed.

Reply:
We have a convention and we are sticking to it.